File: solutions/ruby/matching-brackets/1/matching_brackets.rb
class Brackets
  OPENS  = %w"[ { (".freeze
  CLOSES = %w"] } )".freeze

  def self.paired?(text)
    symbols = []

    text.chars.each do |c|
      symbols << c if OPENS.include?(c)

      if CLOSES.include?(c)
        current = symbols.pop
        index   = CLOSES.index(c)
        return false unless OPENS.index(current) == index
      end
    end

    symbols.empty?
  end
end

File: solutions/javascript/matching-brackets/1/matching-brackets.js
const pairs = {
  '(': ')',
  '[': ']',
  '{': '}'
};

export const isPaired = (text) => {
  let symbols = [];

  for (let i = 0; i < text.length; i++) {
    const current   = text.charAt(i);
    const isOpening = Object.keys  (pairs).includes(current);
    const isClosing = Object.values(pairs).includes(current);

    if (isOpening) {
      symbols.push(current);
    }

    if (isClosing && pairs[symbols.pop()] !== current) {
      return false;
    }
  }

  return symbols.length === 0;
};
